File: src/main/java/domain/freeCell/DoubleFreeCell.java
package domain.freeCell;

import domain.*;
import domain.deal.Deal;
import domain.deal.DealStep;
import domain.deal.ElementTarget;
import domain.deal.Payload;
import domain.deal.steps.DealToFoundation;
import domain.deal.steps.DealToTableau;
import domain.deal.steps.FilterAces;
import domain.deal.steps.FilterByRank;


/**
 * FreeCell with two decks, six reserves.
 */
public class DoubleFreeCell extends FreeCellDomain {

	private Deal deal;
	private Stock stock;
	private Reserve reserve;
	private Tableau tableau;

	/**
	 * Require two decks.
	 */
	@Override
	protected Stock getStock() {
		if (stock == null) { stock = new Stock(2); }
		return stock;
	}


    /**
     * Default Reserve has five Free piles..
     *
     * @return
     */
    protected Reserve getReserve() {
        if (reserve == null) {
            reserve = new Reserve();
            for (int i = 0; i < 6; i++) { reserve.add (new FreePile()); }
        }
        return reserve;
    }

    /** Override deal as needed. */
    @Override
    public Deal getDeal() {
        if (deal == null) {
            deal = new Deal()
                    .append(new FilterAces(4))
                    .append(new DealToFoundation())
                    .append(new DealToTableau(10));
        }

        return deal;
    }

    /**
     * Default Tableau has eight columns
     *
     * @return
     */
    protected Tableau getTableau() {
        if (tableau == null) {
            tableau = new Tableau();
            for (int i = 0; i < 10; i++) { tableau.add (new Column()); }
        }
        return tableau;
    }

	public DoubleFreeCell() {
		super ("DoubleFreeCell");
	}
}

File: src/main/scala/org/combinators/solitaire/freecell/doublefreecell.scala
package org.combinators.solitaire.freecell

import org.combinators.solitaire.domain._

/**
 * Any time you change a twirl template, you HAVE to force a regeneration. Unfortunately, this
 * is accomplished ONLY by launching a web server and connecting via localhost:9000
 */
package object doublefreecell extends variationPoints  {

  override val numDecks:Int = 2

  // can TURN corner from K back to ACE
  override def buildOnFoundation(card: MovingCard.type): Constraint = {
    val topDestination = TopCardOf(Destination)
    OrConstraint(AndConstraint(SameSuit(card, topDestination), IsKing(topDestination), IsAce(card)),
                 AndConstraint(NextRank(card, topDestination),  SameSuit(card, topDestination)))
  }
  override def buildOnFoundationMultiple(cards: MovingCards.type): Constraint = {
    val topDestination = TopCardOf(Destination)
    OrConstraint(AndConstraint(SameSuit(BottomCardOf(cards), topDestination), IsKing(topDestination), IsAce(BottomCardOf(cards))),
                 AndConstraint(NextRank(BottomCardOf(cards), topDestination),  SameSuit(BottomCardOf(cards), topDestination)))
  }
  override val structureMap:Map[ContainerType,Seq[Element]] = Map(
    StockContainer -> Seq.fill[Element](2)(Stock(numDecks)),   // two decks must be present even though not visible
    Tableau -> Seq.fill[Element](10)(Column),        // TEN of these
    Foundation -> Seq.fill[Element](4)(Pile),
    Reserve -> Seq.fill[Element](6)(FreeCellPile),   // SIX of these
  )

  override val layoutMap:Map[ContainerType, Seq[Widget]] = Map (
    Reserve -> horizontalPlacement(10, 10, 6, card_height),
    Foundation -> horizontalPlacement(400+2*card_width, 10, 4, card_height),
    Tableau -> horizontalPlacement(10, 200, num = 10, height = card_height*8),  // estimate
  )

  def getDeal: Seq[DealStep] = {
    var deal:Seq[DealStep] = Seq()
    var colNum = 0
    // Deal cards to all columns (from left to right) until none left, in which case
    // only first four have the extra cards.
    var numDealt = 0
    while (numDealt < 52*numDecks) {
      deal = deal :+ DealStep(ElementTarget(Tableau, colNum), Payload())
      colNum += 1
      numDealt += 1
      if (colNum > 9) { colNum = 0}
    }

    deal
  }

  val doubleFreecell:Solitaire = {

    Solitaire(name="DoubleFreeCell",
      structure = structureMap,
      layout = Layout(layoutMap),
      deal = getDeal,
      /** from element can infer ks.ViewWidget as well as Base Element. */
      specializedElements = Seq(FreeCellPile),

      /** All rules here. */
      moves = Seq(tableauToTableauMove, tableauToFoundationMove, fromTableauToReserve, fromReserveToReserve, fromReserveToTableau, fromReserveToFoundation ),
      // fix winning logic
      logic = BoardState(Map(Foundation -> 104)),
      customizedSetup = Seq.empty,

      solvable = true,
      autoMoves = false   // TOO Complicated, because of issue with multipleaces....
    )
  }

}
